src/field_theory/minpoly.lean

```diff
@@ -386,38 +386,94 @@ by simpa [sub_eq_add_neg] using add_algebra_map hx (-a)
+over the fraction field. Compared to `minpoly.gcd_domain_eq_field_fractions`, this version is useful
+if the element is in a ring that is already a `K`-algebra. -/
+lemma gcd_domain_eq_field_fractions' [algebra K S] [is_scalar_tower R K S] :
+  minpoly K s = (minpoly R s).map (algebra_map R K) :=
+begin
+  let L := fraction_ring S,
+  rw [← gcd_domain_eq_field_fractions K L hs],
+  refine minpoly.eq_of_algebra_map_eq (is_fraction_ring.injective S L)
+    (is_integral_of_is_scalar_tower _ hs) rfl
 end
 
 /-- For GCD domains, the minimal polynomial divides any primitive polynomial that has the integral
 element as root. -/
-lemma gcd_domain_dvd {A R : Type*} (K : Type*)
-  [comm_ring A] [is_domain A] [normalized_gcd_monoid A] [field K]
-  [comm_ring R] [is_domain R] [algebra A K]
-  [is_fraction_ring A K] [algebra K R] [algebra A R] [is_scalar_tower A K R]
-  {x : R} (hx : is_integral A x)
-  {P : A[X]} (hprim : is_primitive P) (hroot : polynomial.aeval x P = 0) :
-  minpoly A x ∣ P :=
+lemma gcd_domain_dvd (hinj : function.injective (algebra_map R S))
+  {P : R[X]} (hP : P ≠ 0) (hroot : polynomial.aeval s P = 0) : minpoly R s ∣ P :=
 begin
-  apply (is_primitive.dvd_iff_fraction_map_dvd_fraction_map K
-    (monic.is_primitive (monic hx)) hprim).2,
-  rw ← gcd_domain_eq_field_fractions K hx,
+  let K := fraction_ring R,
+  let L := fraction_ring S,
+  haveI : no_zero_smul_divisors R L,
+  { refine no_zero_smul_divisors.iff_algebra_map_injective.2 _,
+    rw [is_scalar_tower.algebra_map_eq R S L],
+    exact (is_fraction_ring.injective S L).comp hinj },
+  let P₁ := P.prim_part,
+  suffices : minpoly R s ∣ P₁,
+  { exact dvd_trans this (prim_part_dvd _) },
+  have hP₁ : polynomial.aeval s P₁ = 0,
+  { rw [eq_C_content_mul_prim_part P, map_mul, aeval_C] at hroot,
+    have hcont : P.content ≠ 0 := λ h, hP (content_eq_zero_iff.1 h),
+    replace hcont := function.injective.ne hinj hcont,
+    rw [map_zero] at hcont,
+    exact eq_zero_of_ne_zero_of_mul_left_eq_zero hcont hroot },
+  apply (is_primitive.dvd_iff_fraction_map_dvd_fraction_map K (monic hs).is_primitive
+    P.is_primitive_prim_part).2,
+  let y := algebra_map S L s,
+  have hy : is_integral R y := hs.algebra_map,
+  rw [← gcd_domain_eq_field_fractions K L hs],
   refine dvd _ _ _,
-  rwa ← is_scalar_tower.aeval_apply
+  rw [aeval_map, aeval_def, is_scalar_tower.algebra_map_eq R S L, ← eval₂_map, eval₂_at_apply,
+    eval_map, ← aeval_def, hP₁, map_zero]
+end
+
+lemma gcd_domain_degree_le_of_ne_zero (hinj : function.injective (algebra_map R S))
+  {p : R[X]} (hp0 : p ≠ 0) (hp : polynomial.aeval s p = 0) :
+  degree (minpoly R s) ≤ degree p :=
+begin
+  rw [degree_eq_nat_degree (minpoly.ne_zero hs), degree_eq_nat_degree hp0],
+  norm_cast,
+  exact nat_degree_le_of_dvd (gcd_domain_dvd hs hinj hp0 hp) hp0
+end
+
+omit hs
+
+lemma gcd_domain_unique (hinj : function.injective (algebra_map R S)) {P : R[X]} (hmo : P.monic)
```

Comment:
These are strict generalizations of the existing results for fields (due to [comm_group_with_zero.normalized_gcd_monoid](https://leanprover-community.github.io/mathlib_docs/algebra/gcd_monoid/basic.html#comm_group_with_zero.normalized_gcd_monoid)), right? In that case I'd like to see these replace the existing results (renaming those to `..._aux`).

Reply:
Ah yes, I didn't realize that a field is of course a `normalized_gcd_monoid`.

Reply:
Hmm, we cannot have a single result for the case of fields and `normalized_gcd_domain`, since this one requires `function.injective (algebra_map _ _)`, that is automatic in the case of fields but it's not something can be inferred. Renaming the field case with `..._aux` doesn't seem a good idea to me, since usually this is the version that one wants to use. What do you think?

Reply:
I am stupid, having `no_zero_smul_divisors` solves everything.

Reply:
OK, I have renamed the field lemmas to `..._aux` (they are needed to prove the general version, so we cannot erase them), and deleted the `'` for the gcd version. The only drawback is that the gcd version is not exactly more general, since it does require the element to be in a commutative ring (and this is really needed, since we take its field of fraction), while the field version allows noncommutative rings. Let's see if this requires a lot of modifications.